drivers: eSPI: add eSPI driver support in NPCX7 series

MulinChao opened the pull request:
This PR introduces the Intel Enhanced Serial Peripheral Interface (eSPI) support by three major CLs.

The first one introduces eSPI bus device in npcx7. It includes eSPI device tree declarations, Zephyr eSPI api implementation, OOB (Out of Band tunneled SMBus) support, and configuration files for eSPI test suites in samples/drivers/espi folder.

The second CL contains the drivers of NPCX Host Sub-Modules that serve as an interface between the Host and Core domains. It includes:
- Keyboard and Mouse Controller (KBC) interface.
- Power Management (PM) channels.
- Shared Memory mechanism (SHM).
- Core Access to Host Modules (C2H).
- Mobile System Wake-Up functions (MSWC).
- Serial Port (Legacy UART)

The tasks in the application layer such as 8042, ACPI, and host command can interact with this driver by connecting api or callback functions.

The last one adds a host command peripheral devices for npcx7 series. By add interact api implementation in the hardware bus driver such as ec_host_cmd_periph_npcx_data_received(), the EC host command framework can send and receive data on the npcx7 EC eSPI bus.

Currently, this PR has passed:
- VW handshaking of eSPI test suite.
- OOB transaction through get_pch_temp() function of eSPI test suite.
- port 80 callback function of eSPI test suite.
- KBC IBF/OBE interrupt notification
- ACPI/Host Command (PM Channels) IBF interrupt notification
- EC Host command framework basic test. (We didn't enable the EC_HOST_CMD subsystem by default. Once host sent GET_BOARD_VER command (0x06) request, the main routine in eSPI test suite sent the fixed response data of it directly.)

All the tests are base on Nuvoton host emulator which is an FPGA board and can generate eSPI transactions by the script.

MulinChao committed:
drivers: eSPI: add eSPI driver support for NPCX7 series.

In npcx7 series, all of them support the Intel Enhanced Serial
Peripheral Interface (eSPI) Revision 1.0. This specification provides a
path for migrating host sub-devices via LPC to a lower pin count, higher
bandwidth bus. In addition to Host communication via the peripheral
channel, it provides virtual wires support, out-of-band communication,
and device mastering option over the Chipset SPI flash.

Becisdes introducing eSPI device in npcx7, this CL also includes:

1. Add eSPI device tree declarations.
2. Add npcx7-espi-vws-map.dtsi to present the relationship between eSPI
   Virtual-Wire signals, eSPI registers, and wake-up input sources.
3. Zephyr eSPI api implementation.
4, Add OOB (Out of Band tunneled SMBus) support.
5. Add configuration files for eSPI test suites.

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

MulinChao committed:
drivers: espi: add host sub-modules support for NPCX7 series.

This CL contains the drivers of NPCX Host Sub-Modules that serve as an
interface between the Host and Core domains. For most of them, the Host
can configure these modules via eSPI(Peripheral Channel)/LPC by
accessing 'Configuration and Control register Set' which IO base address
is 0x4E as default. And the interrupts in core domain help handling any
events from host side.

In this commit, we introduced six host sub-modules. It includes:
 1. Keyboard and Mouse Controller (KBC) interface.
 2. Power Management (PM) channels.
 3. Shared Memory mechanism (SHM).
 4. Core Access to Host Modules (C2H).
 5. Mobile System Wake-Up functions (MSWC).
 6. Serial Port (Legacy UART)

The tasks in application layer such as 8042, ACPI and host command can
cooperation with this driver by connecting api or callback functions.

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

MulinChao committed:
drivers: eSPI: fixed build error in eSPI socketpair test suite.

Fixed build error in in eSPI socketpair test suite since wrong function
name for eSPI flash channel api.

This CL also fixed [-Werror=unused-function] warning by adding inline
attribute in case someone includes "espi_utils.h" and doesn't call
espi_manage_callback() function.

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

jettr reviewed (changes requested):
I only had a change to look at the first commit. Will look at the other 3 tomorrow morning.

MulinChao commented on drivers/espi/Kconfig.npcx:
```diff
@@ -0,0 +1,34 @@
+if ESPI_NPCX
+
+config ESPI_OOB_CHANNEL
+	default  y
```
I follow what Intel did in https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/espi/Kconfig.xec#L14. I found 'config ESPI_OOB_CHANNEL' doesn't declare 'default' property in https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/espi/Kconfig#L66. Is it still an overriding case?

jettr replied:
Why providing the test after bool, it makes it overrideable

Oh, you are trying to select the existing Kconfig option. I see. I though you were defining a new options. Can you add a comment saying that we are defaulting the existing option to y since that is the typical use case

MulinChao replied:
Done.

MulinChao commented on drivers/espi/Kconfig.npcx:
```diff
@@ -0,0 +1,34 @@
+config ESPI_PERIPHERAL_UART
+	default y
+
+# Disable only for this board to check notifications
```
Will move to the project configuration file in samples/drivers/espi

MulinChao replied:
Done.

MulinChao commented on drivers/espi/espi_npcx.c:
```diff
@@ -0,0 +1,870 @@
+	/* mapping table between eSPI reset signal and wake-up input */
+	struct npcx_wui espi_rst_wui;
+	/* pinmux configuration */
+	const uint8_t   alts_size;
```
Done.

MulinChao commented on drivers/espi/espi_npcx.c:
```diff
@@ -0,0 +1,870 @@
+
+/* eSPI bus interrupt configuration structure and macro function */
+struct espi_bus_isr {
+	uint32_t status_bit; /* bit in ESPISTS register */
```
It is bit order for 32 bits register. I will replace uint32_t with uint8_t.

MulinChao replied:
Done.

MulinChao commented on drivers/espi/espi_npcx.c:
```diff
@@ -0,0 +1,870 @@
+
+#define NPCX_ESPI_BUS_INT_ITEM(event, isr) {     \
+	.status_bit = NPCX_ESPISTS_##event,      \
+	.int_en_bit = NPCX_ESPIIE_##event##IE,   \
```
I'd like to keep the register bit name as the same as our datasheet. Or it will hard to find it in the datasheet and divergence issue with our HAL.

jettr replied:
This is just putting an underscore between a number and text. Since this is a macro anyway, you won't be able to look up this up in the datasheet since 3IE doesn't exist in code anywhere.

Either way, it is not that big of a deal, so you can leave it the way it is; it just seems like a more consistent name with that extra underscorce

MulinChao replied:
Thanks. What I mean it's the result of this macro is mapping to register bit name in "reg_def.h". (ie. NPCX_ESPIIE_CFGUPDIE -is composed by Chip family + register name + register bit/field name)  If we want to apply this, it also need to modify the register bit name in reg_def.h.

MulinChao commented on drivers/espi/espi_npcx.c:
```diff
@@ -0,0 +1,870 @@
+ * eSPI VW input/Output signal configuration tables. Please refer
+ * npcx7-espi-vws-map.dtsi device tree file for more detail.
+ */
+static const struct npcx_vw_in_config vw_in_tbl[] = {
```
No, it doesn't

MulinChao commented on drivers/espi/espi_npcx.c:
```diff
@@ -0,0 +1,870 @@
+				  .evt_details = 0,
+				  .evt_data = 0 };
+	uint8_t chg_mask = GET_FIELD(inst->ESPICFG, NPCX_ESPICFG_HCHANS_FIELD)
+			 ^ GET_FIELD(inst->ESPICFG, NPCX_ESPICFG_CHANS_FIELD);
```
Done.

MulinChao commented on dts/arm/nuvoton/npcx/npcx7-espi-vws-map.dtsi:
```diff
@@ -0,0 +1,93 @@
+ * |--------------------------------------------------------------------------|
+ * | VW idx | SLV reg | Wire Bit 3   | Wire Bit 2   | Wire Bit 1| Wire Bit 0  |
+ * |--------------------------------------------------------------------------|
+ * |                 Input (Master-to-Slave) Virtual Wires                    |
```
I still suggest using Master-to-Slave which also is indicated in our datasheet. The input signal means this signal is sent by bus master (PCH) to bus slave, (EC)

MulinChao commented on soc/arm/nuvoton_npcx/common/reg/reg_def.h:
```diff
@@ -308,4 +308,163 @@ struct gpio_reg {
+};
+
+/* eSPI register fields */
+#define NPCX_ESPICFG_PCHANEN             0
```
For register bit definition, we only declare bit order in reg_def.h and use BIT(x) in driver instance. I'd like to keep it for consistency. If it's crucial feature must meet, I suggest to create a refactor PR after all npcx driver land.

jettr replied:
We can keep it for consistent. Just a small comment

MulinChao commented on soc/arm/nuvoton_npcx/common/soc_dt.h:
```diff
@@ -18,6 +18,27 @@
+#define DT_NPCX_CLK_CFG_ITEMS_LEN(inst) DT_INST_PROP_LEN(inst, clocks)
+
+/* Macro function to construct a list of npcx_clk_cfg items by UTIL_LISTIFY */
+#define DT_NPCX_CLK_CFG_ITEMS_FUC(idx, inst) \
```
It extends an array defined from clocks property in a device tree file by UTIL_LISTIFY macro. Please refer https://github.com/zephyrproject-rtos/zephyr/blob/master/include/sys/util.h#L617 for more detail.

MulinChao replied:
I submitted a CL to add more detail and explanations for all macro functions in this PR. Please take a look.

MulinChao commented on soc/arm/nuvoton_npcx/common/soc_dt.h:
```diff
@@ -99,4 +120,48 @@
+/* Construct a npcx_wui structure from wui_map property of vw table by name */
+#define DT_NPCX_VW_WUI_ITEM(name)			                       \
+	{                                                                      \
+	  .table = DT_PROP(DT_PHANDLE(DT_PHANDLE_VW_WUI(name), miwus),  index),\
```
The index is the property of miwu device defined in device tree file. Please refer https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/interrupt-controller/nuvoton%2Cnpcx-miwu.yaml#L14

MulinChao commented on soc/arm/nuvoton_npcx/common/soc_dt.h:
```diff
@@ -99,4 +120,48 @@
+#define DT_NPCX_ESPI_WUI_ITEM(name)				               \
+	{                                                                      \
+	  .table = DT_PROP(DT_PHANDLE(DT_PHANDLE_FROM_WUI_NAME(name),          \
+					miwus), index),                        \
```
Please see the above.

MulinChao commented on drivers/espi/espi_utils.h:
```diff
@@ -20,7 +20,7 @@
  * @return 0 on success, negative errno otherwise.
  */
-static int espi_manage_callback(sys_slist_t *callbacks,
+static inline int espi_manage_callback(sys_slist_t *callbacks,
```
No, it won't since the larger CL used this function. But I noticed this PR cannot pass the "socketpair" test suite. The root cause is wrong function names for eSPI flash APIs. Hence, I added the fix and raised up the order of this CL. I'm also curious why the other eSPI driver CL won't encounter this problem?

jettr replied:
I think this should be the parent of your larger CL. It can stand on its own and fixes a potential problem. That way your larger eSPI commit won't have any test failures.

MulinChao replied:
Done.

MulinChao replied:
Done.

MulinChao committed:
soc: npcx7: add comments for macro funcs used for device tree files.

This CL adds more comments for each macro functions used for device tree
file for better explanations. It also changes all hex values in soc.c to
lower case.

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

MulinChao commented on soc/arm/nuvoton_npcx/common/soc_dt.h:
```diff
@@ -18,6 +18,27 @@
 	  .bit  = DT_PHA(DT_DRV_INST(inst), clocks, bit),                      \
 	}
 
+/* Construct a npcx_clk_cfg structure from clocks property at index 'i' */
```
Simon also reminded me in the other PR. In the beginning, I think it's unnecessary because they are only used for npcx series internally. But it seems to give reviewers a headache without detailed comment. I will add a CL to add comments to describe parameter, return type, and usage of each macro function.

MulinChao replied:
Done.

albertofloyd commented on drivers/espi/espi_handlers.c:
```diff
@@ -55,7 +55,7 @@ static inline int z_vrfy_espi_write_lpc_request(const struct device *dev,
 	Z_OOPS(z_user_from_copy(&data_copy, data, sizeof(*data)));
 
-	return z_impl_espi_write_lpc_request(dev, op, data_copy);
+	return z_impl_espi_write_lpc_request(dev, op, &data_copy);
```
This looks correct as is now, no need to changed since recently fixed by @ceolin

MulinChao replied:
Ok. I will get rid of this line.

MulinChao replied:
Done.

MulinChao replied:
No, it still doesn't work and block this PR. Please refer https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/espi/espi_handlers.c#L58. Might I ask which PR fix it? If not, I will add it back in this CL.

albertofloyd commented on samples/drivers/espi/prj_npcx7m6fb_evb.conf:
```diff
@@ -0,0 +1,15 @@
+CONFIG_LOG=y
+CONFIG_LOG_BUFFER_SIZE=4096
+CONFIG_LOG_PROCESS_THREAD_SLEEP_MS=100
+<<<<<<< Updated upstream
```
Unmerged problem

MulinChao replied:
Oops. I fixed it in the other CL. Will change the order of this.

albertofloyd commented on drivers/espi/Kconfig:
```diff
@@ -112,6 +112,12 @@ config ESPI_PERIPHERAL_DEBUG_PORT_80
 	help
 	  Enables debug Port 80 over eSPI peripheral channel.
 
+config ESPI_PERIPHERAL_EC_HOST_CMD
```
Is this different than CONFIG_ESPI_PERIPHERAL_HOST_IO?

MulinChao replied:
No, this config is to enable the support of 'EC host command framework'. Please https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ec_host_cmd_periph/Kconfig#L6 for more detail.

MulinChao replied:
Done

jettr commented on drivers/ec_host_cmd_periph/ec_host_cmd_npcx.c:
```diff
@@ -0,0 +1,81 @@
+ * Copyright (c) 2020 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
This should be an eSPI driver that works for all eSPI implementations (versus a npcx implementation). NPCX should have an eSPI implementation, but the host cmd periph should be based on a bus like eSPI, SPI, or I2C.

I also believe that I need to update the ec_host_cmd_periph API to include a blocking `read` API function. I used semaphores in the initial API instead, but now that I am implementing the `SPI` version of this periph driver, I realize it make things more complicated than it needs to be.

I have a *very* rough draft of the `SPI` implementation for this driver that I [uploaded](https://github.com/jettr/zephyr/tree/spi_periph_drat) to show you what I mean.

MulinChao replied:
That's my original plan, too. eSPI driver offers  espi_api_lpc_read_request, espi_api_lpc_write_request  and callbacks to handle the events from host sub-modules such as KBC, APCI IO ports and so on. The Chromium host cmd framework needs a event from the IO port and handle massive data via shared memory or mailbox. 

But the reality is eSPI api functions only support 8041 so far. Please refer  https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/espi.h#L212 . We need to add more op codes after the interface between host commands and espi peripheral is finalized. In order to test the hardware behavior on npcx7 first, this CL follows what you did fo the simulator and passed the sanity test on my side 

Should I drop this commit in this PR? After all host_cmd_periph_xxx api functions are landed, we can create the other PR to support host command framework.

jettr replied:
Let's drop the host command peripheral for now then and we can add it when eSPI and the host command framework are both in a better state

MulinChao replied:
Got it. I will drop it in this PR.

MulinChao replied:
Done.

jettr commented on samples/drivers/espi/prj_npcx7m6fb_evb.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_LOG_PROCESS_THREAD_SLEEP_MS=100
+
+# EC Host Command Driver
+CONFIG_ESPI_PERIPHERAL_EC_HOST_CMD=y
```
We shouldn't have this config option yet (since it isn't implement yet, right)?
```suggestion
```

MulinChao replied:
Oops. I miss this line when I dropped the implementation CL. Thanks.

MulinChao replied:
Done.

jettr reviewed (commented):
this looks good to me (with minor comment). I have request that other also review this PR. I also know that we will need to wait until the 2.5 merge window open to land this.

sjg20 commented on soc/arm/nuvoton_npcx/npcx7/Kconfig.defconfig.series:
```diff
@@ -35,6 +35,12 @@ config GPIO_NPCX
+	default y
+	depends on ESPI
+	help
+	  Enable support for NPCX ESPI driver.
```
This doesn't help much. Ideas to expand this:

- some mention of what eSPI is for
- what features the driver supports
- link to eSPI spec

MulinChao replied:
Okay, I know eSPI is not a peripheral interface well known. I will add its spec link of Interl website.

MulinChao replied:
Done.

sjg20 commented on soc/arm/nuvoton_npcx/npcx7/soc.c:
```diff
@@ -56,3 +56,46 @@ NPCX_REG_OFFSET_CHECK(espi_reg, OOBTXBUF, 0x380);
 NPCX_REG_OFFSET_CHECK(espi_reg, OOBCTL_DIRECT, 0x3FC);
```
Can we use lower-case hex everywhere?

MulinChao replied:
Done.

galak commented on dts/bindings/espi/nuvoton,npcx-espi.yaml:
```diff
@@ -0,0 +1,33 @@
+            For example the WUI mapping on NPCX7 would be
+               espi_rst_wui = <&wui_cr_sin1>;
+
+        required: true
```
got `required: true` twice here.

MulinChao replied:
Thanks for reminding. I will update it.

MulinChao replied:
Done.
